design/0000-launcher.md

```diff
@@ -0,0 +1,121 @@
+        }),
+    )
+    defer lnchr.Shutdown()
+}
+```
+
+### Configuration Options
+
+| Config Option               | Env Variable                        | Required | Default              |
+| --------------------------  | ----------------------------------- | -------- | -------------------- |
+| WithServiceName             | OTEL_SERVICE_NAME                   | y        | -                    |
+| WithServiceVersion          | OTEL_SERVICE_VERSION                | n        | -                    |
+| WithHeaders                 | OTEL_EXPORTER_OTLP_HEADERS          | n        | {}                   |
+| WithTracesExporterEndpoint  | OTEL_EXPORTER_OTLP_TRACES_ENDPOINT  | n        | localhost:4317       |
+| WithTracesExporterInsecure  | OTEL_EXPORTER_OTLP_TRACES_INSECURE  | n        | false                |
+| WithMetricsExporterEndpoint | OTEL_EXPORTER_OTLP_METRICS_ENDPOINT | n        | localhost:4317       |
+| WithMetricsExporterInsecure | OTEL_EXPORTER_OTLP_METRICS_INSECURE | n        | false                |
+| WithLogLevel                | OTEL_LOG_LEVEL                      | n        | info                 |
+| WithPropagators             | OTEL_PROPAGATORS                    | n        | tracecontext,baggage |
+| WithResourceAttributes      | OTEL_RESOURCE_ATTRIBUTES            | n        | -                    |
+| WithMetricsReportingPeriod  | OTEL_EXPORTER_OTLP_METRICS_PERIOD   | n        | 30s                  |
+| WithMetricsEnabled          | OTEL_METRICS_ENABLED                | n        | true                 |
+| WithTracesEnabled           | OTEL_TRACES_ENABLED                 | n        | true                 |
+| WithProtocol                | OTEL_EXPORTER_OTLP_PROTOCOL         | n        | grpc                 |
+
+### Additional Configuration Options
+
+Not yet implemented but still desired configuration options would include:
+
+- `OTEL_EXPORTER_OTLP_PROTOCOL`: `http/protobuf`
+    - ideally, selecting this would change the default endpoint to localhost:4318
+
+### Using a Vendor-Specific Package
+
+Vendors may create and maintain convenience configuration packages to more easily setup for export to a specific backend. For example, using a Honeycomb package would look like this:
+
+Minimal setup, which sends to the Honeycomb endpoint and requires `HONEYCOMB_API_KEY` environment variable:
+
+```go
+import (
+    "github.com/opentelemetry-go-contrib/otel-launcher-go/launcher"
+    _ "github.com/honeycombio/otel-launcher-go/launcher/honeycomb"
+)
+
+func main() {
+    lnchr, err := launcher.ConfigureOpentelemetry()
+    defer lnchr.Shutdown()
+}
+```
+
+Alternatively, to set the Honeycomb API key directly in code:
+
+```go
+import (
+    "github.com/opentelemetry-go-contrib/otel-launcher-go/launcher"
+    "github.com/honeycombio/otel-launcher-go/launcher/honeycomb"
+)
+
+func main() {
+    lnchr, err := launcher.ConfigureOpentelemetry(
+        honeycomb.WithApiKey("api-key"),
```

Comment:
I assume this would return a `launcher.Option`, probably by wrapping `launcher.WithHeaders()` like this:

```go
func WithAPIKey(key string) launcher.Option {
  return launcher.WithHeaders(map[string]string{"HONEYCOMB_API_KEY": key})
}
```

Would the `launcher.Option` type operate on an exported `launcher.Config` type allowing the implementation of wholly-new option implementations, or would everything need to be done in terms of existing `launcher.Option` implementations?  We've traditionally not exported the config type to avoid unintended/unsupportable behavior, but this might be a place where doing it make sense.

Reply:
This assumption is correct, that it would return a `launcher.Option`.

Regarding unintended/unsupportable behavior, vendors can add config validation for their own use cases. For example, our [vendor package (WIP)](https://github.com/honeycombio/otel-launcher-go/blob/mike/reorganise/honeycomb/honeycomb.go#L45) has a `launcher.ValidateConfig` that returns errors for invalid headers.